### PR TITLE
[DO NOT REVIEW] Subgraph API

### PR DESCRIPTION
**Pull request by reminisce**
## Description ##
Incremental change to the dev branch. Not ready for review.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/executor/graph_executor.cc [hunk 1]
```diff
@@ -1699,6 +1701,146 @@ GraphExecutor::CachedSegOpr GraphExecutor::CreateCachedSegOpr(size_t topo_start,
...
+                                   const std::map<std::string, Context>& ctx_map,
+                                   const std::vector<Context>& in_arg_ctxes,
+                                   const std::vector<Context>& aux_state_ctxes) {
+  auto subgraph_prop = op::SubgraphPropertyRegistry::Get()->CreateSubgraphProperty(prop_name);
+  nnvm::Symbol ret = src.Copy();
+  nnvm::Graph g;
+  g.outputs = ret.outputs;
+  g = InferForwardAttrs(g, arg_shapes, arg_dtypes, arg_stypes, default_ctx,
+                        ctx_map, in_arg_ctxes, aux_state_ctxes);
+  subgraph_prop->SetAttr("graph", g);
```
Comment (zheng-da) [inline src/executor/graph_executor.cc]: This interface doesn't allow customizing `g`. I think it's better to provide an interface such as `set_graph` and allow the child class to override this virtual method.

File: src/executor/graph_executor.cc [hunk 2]
```diff
@@ -1718,6 +1860,11 @@ Executor *Executor::SimpleBind(nnvm::Symbol symbol,
                                std::unordered_map<std::string, NDArray>* shared_buffer,
                                Executor* shared_exec) {
   auto exec = new exec::GraphExecutor();
+  if (!exec->subgraph_property().empty()) {
+    symbol = exec::PartitionGraph(symbol, exec->subgraph_property(), arg_shape_map, arg_dtype_map,
```
Comment (ZhennanQin) [inline src/executor/graph_executor.cc]: Many subgraph pass can only be used on inference, eg. default_subgraph_op doesn't have FGradient attribute and can't support backward computation. Maybe default_subgraph_op will support backward in future, but some mkldnn fusion pass is designed to be used in inference only. Shall we allow subgraph_property to describe if it's inference only and apply it as it requires?

**Comment by ZhennanQin** [comment]
Can this change cover hybridize gluon model? I don't see the change in cached_op.

**Comment by ZhennanQin** [comment]
Another question: what's the sequence should we execute below 2 kinds of subgraph pass:
1. the pass to cover all operators that backends support.(just like default_subgraph_op)
2. Fusion pass to combine several ops into a new ops.

I'm asking this is because, if we want to apply 2 after 1, then we need to have the ability to traverse into subgraph node to do fusion. Do we support this?

File: src/executor/graph_executor.cc (hunk 2 above)
Comment (reminisce) [reply src/executor/graph_executor.cc]: Is it necessary to indicate whether a subgraph property is for inference only or not?

**Comment by reminisce** [comment]
@ZhennanQin CachedOp will be submitted through a another PR. We want to make this PR as small as possible.

I think what you described about fusion can be done in graph partitioning as well. In `CreateSubgraphNode()`, you can replace the subgraph symbol with a new subgraph symbol consisting of fused operators. Does this make sense?

File: src/executor/graph_executor.cc (hunk 1 above)
Comment (reminisce) [reply src/executor/graph_executor.cc]: I think this can be done in `CreateSubgraphSelector()`?

File: src/executor/graph_executor.cc (hunk 2 above)
Comment (ZhennanQin) [reply src/executor/graph_executor.cc]: Yes, that's also my question. We indeed have some fusion passes only for inference, we need a way to disable it on training. If we can't describe this in subgraph property, then we need to find another way to indicate that.

**Comment by ZhennanQin** [comment]
@reminisce Then we need to call PartitionGraph() inside CreateSubgraphNode() to traverse nodes, otherwise we have to implement a separate traverse method. If that meet your expectation, then I'm OK with this solution.

File: src/executor/graph_executor.cc (hunk 1 above)
Comment (zheng-da) [reply src/executor/graph_executor.cc]: Subgraph selector is created by each node. I feel it's more intuitive to customize `g` somewhere subgraph property is created.

**Comment by reminisce** [comment]
@ZhennanQin Yes, I fine with that.

**Pull request merged**